go-controller/pkg/config/config.go

```diff
@@ -1,105 +1,589 @@
+
+	if defaults.K8sToken {
+		token, err := getOVSExternalID("k8s-api-token")
+		if err != nil {
+			logrus.Debugf("Failed to get default Kubernetes API token: %v", err)
+		} else if cfg.Kubernetes.Token == "" {
+			cfg.Kubernetes.Token = token
+		}
 	}
-	if cfg.Default.CniConfPath != "" {
-		CniConfPath = cfg.Default.CniConfPath
+}
+
+func validateKubernetesConfig(cli, file *config) error {
+	Kubernetes.Kubeconfig = getConfigStr(cli.Kubernetes.Kubeconfig, file.Kubernetes.Kubeconfig, "")
+	if Kubernetes.Kubeconfig != "" && !pathExists(Kubernetes.Kubeconfig) {
+		return fmt.Errorf("kubernetes kubeconfig file %q not found", Kubernetes.Kubeconfig)
 	}
-	if cfg.Default.CniPlugin != "" {
-		CniPlugin = cfg.Default.CniPlugin
+	Kubernetes.CACert = getConfigStr(cli.Kubernetes.CACert, file.Kubernetes.CACert, "")
+	if Kubernetes.CACert != "" && !pathExists(Kubernetes.CACert) {
+		return fmt.Errorf("kubernetes CA certificate file %q not found", Kubernetes.CACert)
 	}
-	if cfg.Default.PrivateKey != "" {
-		NbctlPrivateKey = cfg.Default.PrivateKey
+	Kubernetes.Token = getConfigStr(cli.Kubernetes.Token, file.Kubernetes.Token, "")
+	Kubernetes.APIServer = getConfigStr(cli.Kubernetes.APIServer, file.Kubernetes.APIServer, "http://localhost:8443")
+	url, err := url.Parse(Kubernetes.APIServer)
+	if err != nil {
+		return fmt.Errorf("kubernetes API server address %q invalid: %v", Kubernetes.APIServer, err)
+	} else if url.Scheme != "https" && url.Scheme != "http" {
+		return fmt.Errorf("kubernetes API server URL scheme %q invalid", url.Scheme)
 	}
-	if cfg.Default.Certificate != "" {
-		NbctlCertificate = cfg.Default.Certificate
+
+	if strings.HasPrefix(Kubernetes.APIServer, "https") && Kubernetes.CACert == "" {
+		return fmt.Errorf("kubernetes API server %q scheme requires a CA certificate", Kubernetes.APIServer)
 	}
-	if cfg.Default.CACert != "" {
-		NbctlCACert = cfg.Default.CACert
+	return nil
+}
+
+// InitConfig reads the config file and common command-line options and
+// constructs the global config object from them.
+func InitConfig(ctx *cli.Context, defaults *Defaults) error {
+	var cfg config
+	var err error
+	var f *os.File
+
+	// Error parsing a user-provided config file is a hard error
+	configFile := ctx.String("config-file")
+	if configFile != "" {
+		f, err = os.Open(configFile)
+		if err != nil {
+			return fmt.Errorf("failed to open config file %s: %v", configFile, err)
+		}
+	} else {
+		// Failure to find a default config file is not a hard error
+		f, _ = os.Open("/etc/openvswitch/ovn_k8s.conf")
+	}
+	if f != nil {
+		defer f.Close()
+
+		// Parse ovn-k8s config file.
+		if err = gcfg.ReadInto(&cfg, f); err != nil {
+			return fmt.Errorf("failed to parse config file %s: %v", f.Name(), err)
+		}
+		logrus.Infof("Parsed config file %s", f.Name())
+		logrus.Infof("Parsed config: %+v", cfg)
+
+		if defaults != nil {
+			fillDefaults(&cfg, defaults)
+		}
+	}
+
+	// command-line > config file > default values
```

Comment:
I think using global variables for the default values would be easier to read and understand in my opinion and all default values will be in only one place. Or at least a structure with all the default values defined would work as well.

Reply:
Probably a complete example file with all the options listed would be useful.

Reply:
Copy from the PR description itself above and put it somewhere along the README

Reply:
@alinbalutoiu this only works for a couple of the simple structs; for others (Kubernetes and OvnAuth) we need to construct the defaults based on things like whether the address is HTTP or HTTPS, or we need to grab the default value from OVS external-ids.

I'll do some reworking to make as many defaults explicitly specified as we can.

Reply:
@dcbw I was just thinking of having the default values defined as constants, for example:
```
Default.MTU = getConfigInt(cliConfig.Default.MTU, cfg.Default.MTU, 1400)
```
Instead of having that 1400 over there to be defined as a constant in the beginning of the file.
Anyway that is just a thought, the way you did it works as well.